### DISCUSSION TITLE
question/implement-basic-throttle

### DISCUSSION THREAD

**Discussion by giscus** [us-fe/us-fe.github.io#19] [Announcements]
# question/implement-basic-throttle

北美前端面试攻略

https://us-fe.github.io/question/implement-basic-throttle.html

**Comment by xiaoyu2er** [comment]
这个题目首先要理解 防抖（debounce） 和 节流（throttle） 分别是什么，区别是什么

https://towardsdev.com/debouncing-and-throttling-in-javascript-8862efe2b563


> Debouncing
> + Debouncing is a programming practice used to ensure that time-consuming tasks do not fire so often, that it stalls the performance of the web page. The debounced function will ignore all calls to it until the calls have stopped for a specified time period. Only then will it call the original function.
> + Debouncing forces a function to wait a certain amount of time before running again. In other words, it limits the rate at which a function gets invoked.

> Throttling
> + Throttling or sometimes is also called throttle function is a practice used in websites. To throttle a function means to ensure that the function is called at most once in a specified time period (for instance, once every 10 seconds). This means throttling will prevent a function from running if it has run “recently”. Throttling also ensures a function is run regularly at a fixed rate.
> + Throttling is used to call a function after every millisecond or a particular interval of time only the first click is executed immediately.

https://segmentfault.com/a/1190000014292298

> 节流 throttle 与 去抖 debounce的区别主要在触发时机上：
> + debounce(func, wait, options)：创建并返回函数的防反跳版本，将延迟函数的执行（真正的执行）在函数最后一次调用时刻的wait毫秒之后，对于必须在一些输入（多是一些用户操作）停止之后再执行的行为有帮助。将一个连续的调用归为一个，如果连续在wait毫秒内调用，最后只有最后一次会执行
> + throttle(func, wait, options)：创建并返回一个像节流阀一样的函数，当重复调用函数的时候，最多每隔指定的wait毫秒调用一次该函数；不允许方法在每wait毫秒间执行超过一次，如果连续在wait毫秒内调用，最后执行会均匀分布在大约每wait一次

**Comment by xiaoyu2er** [comment]
我们来看一下第一个答案

```js

/**
 * @param {Function} func
 * @param {number} wait
 */
function throttle(func, wait) {
  let waiting = false;
  let lastArgs;
  return function () {
     if (!waiting) {
       waiting = true;
       func.apply(this, arguments);
       setTimeout(() => {
         waiting = false;
         if (lastArgs) func.apply(this, lastArgs);
       }, wait);
     } else {
       lastArgs = [...arguments];
     }
  }
}
```

这个答案可以通过 test case 但主要原因是因为 test 过于简单，没有覆盖题目中出现的复杂情况
```js
// 调用顺序
// -0-1-2-3-4-5-6-7-8-9-0-1-2-3-
// ─A─B─C─ ─D─ ─ ─ ─ ─ ─E─ ─F─G
// 预期执行顺序
// ─A─ ─ ─C─ ─ ─D─ ─ ─ ─E─ ─ ─G
```

```js
const throttled = throttle(func, 300)
throttled('A');

setTimeout(() => {
  throttled('B');
}, 100);

setTimeout(() => {
  throttled('C');
}, 200);

setTimeout(() => {
  throttled('D');
}, 400);

setTimeout(() => {
  throttled('E');
}, 1000);

setTimeout(() => {
  throttled('F');
}, 1200);

setTimeout(() => {
  throttled('G');
}, 1300);

setTimeout(() => {
  console.log(calls);
}, 2000);

// 输出 ['A@0', 'C@303', 'D@401', 'C@706', 'E@1006', 'G@1307']
```
那么问题出在那里呢？

```js
// 调用顺序
// -0-1-2-3-4-5-6-7-8-9-0-1-2-3-
// ─A─B─C─ ─D─ ─ ─ ─ ─ ─E─ ─F─G
// 预期执行顺序
// ─A─ ─ ─C─ ─ ─D─ ─ ─ ─E─ ─ ─G
// 实际的的执行顺序
// ─A─ ─ ─C─D─ ─ ─ ─ ─ ─E─ ─ ─G
```

可以看出是 C D  之间出了问题，原因在于函数只考虑了立即执行完之后的一次执行情况，没有考虑到再往后的执行情况。

**Comment by xiaoyu2er** [comment]
我们再来看一下第二个优化后的答案

```js
function throttle(func, wait) {
  let waiting = false, lastArgs = null;
  return function(...args){
    if(!waiting){
      func.apply(this, args)
      waiting = true;
      let timeout = () => setTimeout(() => {
        waiting = false;
        if(lastArgs){
          func.apply(this, lastArgs);
          waiting = true;
          lastArgs = null;
          timeout();
        }
      }, wait);
      timeout();
    }
    else 
      lastArgs = args
  }
}
```

```js
const calls = [];
const zero = + new Date();
const func = (arg) => {
     calls.push(`${arg}@` + `${ + new Date() - zero }`.slice(-4));
}

const throttled = throttle(func, 300)
// 调用顺序
// -0-1-2-3-4-5-6-7-8-9-0-1-2-3-
// ─A─B─C─ ─D─ ─ ─ ─ ─ ─E─ ─F─G
// 预期执行顺序
// ─A─ ─ ─C─ ─ ─D─ ─ ─ ─E─ ─ ─G

throttled('A');

setTimeout(() => {
  throttled('B');
}, 100);

setTimeout(() => {
  throttled('C');
}, 200);

setTimeout(() => {
  throttled('D');
}, 400);

setTimeout(() => {
  throttled('E');
}, 1000);

setTimeout(() => {
  throttled('F');
}, 1200);

setTimeout(() => {
  throttled('G');
}, 1300);

setTimeout(() => {
  console.log(calls);
}, 2000);

// 输出 ['A@0', 'C@302', 'D@604', 'E@1001', 'G@1302']
// 调用顺序
// -0-1-2-3-4-5-6-7-8-9-0-1-2-3-
// ─A─B─C─ ─D─ ─ ─ ─ ─ ─E─ ─F─G
// 预期执行顺序
// ─A─ ─ ─C─ ─ ─D─ ─ ─ ─E─ ─ ─G
// 实际执行顺序
// ─A─ ─ ─C─ ─ ─D─ ─ ─ ─E─ ─ ─G
```

这里最重要的修改是，每次执行完，都要保证后一次要在 wait 之后被调用，所以在函数内部，主动调用了 timeout 函数来延迟执行下一次调用。
